bin/ch/ch.cpp

```diff
@@ -584,6 +599,108 @@ static HRESULT CreateRuntime(JsRuntimeHandle *runtime)
+
+Error:
+    return hr;
+}
+
+HRESULT CreateParserState(LPCSTR fileContents, size_t fileLength, JsFinalizeCallback fileContentsFinalizeCallback, LPCWSTR fullPath)
+{
+    HRESULT hr = S_OK;
+    HANDLE fileHandle = nullptr;
+    JsValueRef parserStateBuffer = nullptr;
+    BYTE *buffer = nullptr;
+    unsigned int bufferSize = 0;
+
+    IfFailedGoLabel(GetParserStateBuffer(fileContents, fileContentsFinalizeCallback, &parserStateBuffer), Error);
+    IfJsErrorFailLog(ChakraRTInterface::JsGetArrayBufferStorage(parserStateBuffer, &buffer, &bufferSize));
+
+    fileHandle = GetFileHandle(fullPath);
+    IfFalseGo(fileHandle != INVALID_HANDLE_VALUE && fileHandle != nullptr);
+
+    for (unsigned int i = 0; i < bufferSize; i++)
+    {
+        const unsigned int BYTES_PER_LINE = 32;
+        DWORD written = 0;
+        char scratch[3];
+        auto scratchLen = sizeof(scratch);
+        int num = _snprintf_s(scratch, scratchLen, _countof(scratch), "%02X", buffer[i]);
+        Assert(num == 2);
+        IfFalseGo(WriteFile(fileHandle, scratch, (DWORD)(scratchLen - 1), &written, nullptr));
+
+        // Add line breaks so this block can be readable
+        if (i % BYTES_PER_LINE == (BYTES_PER_LINE - 1) && i < bufferSize - 1)
+        {
+            IfFalseGo(WriteFile(fileHandle, "\n", 1, &written, nullptr));
+        }
+    }
+
+Error:
+    if (fileHandle != nullptr)
+    {
+        CloseHandle(fileHandle);
+    }
+    return hr;
+}
+
+HRESULT CreateParserStateAndRunScript(const char* fileName, LPCSTR fileContents, size_t fileLength, JsFinalizeCallback fileContentsFinalizeCallback, char *fullPath)
+{
+    HRESULT hr = S_OK;
+    JsRuntimeHandle runtime = JS_INVALID_RUNTIME_HANDLE;
+    JsContextRef context = JS_INVALID_REFERENCE, current = JS_INVALID_REFERENCE;
+    JsValueRef bufferVal;
+
+    // We don't want this to free fileContents when it completes, so the finalizeCallback is nullptr
+    IfFailedGoLabel(GetParserStateBuffer(fileContents, nullptr, &bufferVal), ErrorRunFinalize);
+
+    // Bytecode buffer is created in one runtime and will be executed on different runtime.
+    IfFailedGoLabel(CreateRuntime(&runtime), ErrorRunFinalize);
+    chRuntime = runtime;
+
+    IfJsErrorFailLogLabel(ChakraRTInterface::JsCreateContext(runtime, &context), ErrorRunFinalize);
+    IfJsErrorFailLogLabel(ChakraRTInterface::JsGetCurrentContext(&current), ErrorRunFinalize);
+    IfJsErrorFailLogLabel(ChakraRTInterface::JsSetCurrentContext(context), ErrorRunFinalize);
+
+    // Initialized the WScript object on the new context
+    if (!WScriptJsrt::Initialize())
+    {
+        IfFailedGoLabel(E_FAIL, ErrorRunFinalize);
+    }
+
+    // This is our last call to use fileContents, so pass in the finalizeCallback
+    IfFailGo(RunScript(fileName, fileContents, fileLength, fileContentsFinalizeCallback, nullptr, fullPath, bufferVal));
+
+    if (false)
+    {
+    ErrorRunFinalize:
```

Comment:
Yeah, this is a weird pattern but it is lifted from `CreateAndRunSerializedScript` and I'm a little resistant to change it since it does work. I'll see if I can merge these two methods and clean up the logic here in the process.